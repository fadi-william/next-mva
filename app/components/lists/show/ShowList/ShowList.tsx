// The imported libs.
import { observer } from "mobx-react";
import * as React from "react";

// The component's model type.
import { TShowList } from "../../../../models/show/ShowList";

// The external imported components.
import ShowListItem from "../ShowListItem/ShowListItem";

// The ShowList's props interface.
interface IShowListProps {
    showList: TShowList;
}

@observer
export default class ShowList extends React.Component<IShowListProps> {

    public render() {
        const { shows, isLoading } = this.props.showList;

        return (
            <div>
                <h1>Shows</h1>
                {!isLoading && <ul>
                    {shows.map((show, idx) => (
                        <ShowListItem key={idx} show={show} />
                    ))}
                </ul>}
                {isLoading && <div>
                    Loading...
                </div>}
            </div>
        );
    }
}
