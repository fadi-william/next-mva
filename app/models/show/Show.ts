// The imported libs.
import { applySnapshot, flow, types } from "mobx-state-tree";
import { getShow } from "../../service/showService";

export type TShow = typeof Show.Type;

// The module's global variables.
let store: TShow = null;

// If the id changed, we will re-initialize the store *(on the client-side).
let currentShowId: number = null;

export const Show = types.model({
    id: types.maybe(types.identifier(types.number)),
    image: types.maybe(types.string),
    isLoading: types.optional(types.maybe(types.boolean), true),
    name: types.maybe(types.string),
    summary: types.maybe(types.string),
})
.actions( (self) => ({
    doneLoading() {
        self.isLoading = false;
    },
}))
.actions( (self) => ({
    getShow: flow(function *() {
        const response = yield getShow(self.id);

        if (response.data) {
            const show = Object.assign({}, response.data);

            show.id = Number(response.data.id);
            show.image = response.data.image.medium;

            applySnapshot(self, show);

            self.doneLoading();
        }
    }),
}))
.actions( (self) => ({
    afterCreate() {
        self.getShow();
    },
}));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(isServer, showId, snapshot = null) {
    // If this is the server.
    if (isServer) {
      currentShowId = showId;
      store = Show.create({
          id: showId,
      });
    }
    if (store === null || currentShowId !== showId) {
        currentShowId = showId;
        store = Show.create({
            id: showId,
        });
    }
    if (snapshot) {
      applySnapshot(store, snapshot);
    }
    return store;
}
