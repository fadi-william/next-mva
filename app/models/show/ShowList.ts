// The imported libs.
import { applySnapshot, types  } from "mobx-state-tree";
import "rxjs/add/operator/debounceTime";
import "rxjs/add/operator/map";
import { Subject } from "rxjs/Subject";

import { getShows } from "../../service/showService";
import { Show } from "./Show";

export type TShowList = typeof ShowList.Type;

// The module's global variables.
let store: TShowList = null;
const getShowsSubject$ = new Subject<void>();

export const ShowList = types.model({
    isLoading: types.optional(types.maybe(types.boolean), true),
    shows: types.optional(types.array(Show), []),
})
// The shows model's related actions.
.actions( (self) => ({
    doneLoading() {
        self.isLoading = false;
    },
    getShows() {
        // Get the latest shows.
        getShowsSubject$.next();
    },
}))
.actions( (self) => ({
    afterCreate() {
        getShowsSubject$
            .debounceTime(200)
            .map(() => {
                return getShows();
            })
            .subscribe((showsService) => {
                showsService.then((showResponse) => {
                    const shows = showResponse.data.map( (showObj) => {
                        return {
                            id: showObj.show.id,
                            name: showObj.show.name,
                        };
                    });

                    applySnapshot(self.shows, shows);

                    if (self.isLoading) {
                        self.doneLoading();
                    }
                });
            });

        // Trigger get shows.
        self.getShows();
    },
    beforeDestroy() {
        getShowsSubject$.unsubscribe();
    },
}))
// Derived data from the shows model.
.views( (self) => ({
    get totalShows() {
        return self.shows.length;
    },
}));

// Utility function - Instantiates a part of the mobx-state-tree store.
export function getStore(isServer, snapshot = null) {
    // If this is the server.
    if (isServer) {
      store = ShowList.create();
    }
    if (store === null) {
        store = ShowList.create();
    }
    if (snapshot) {
      applySnapshot(store, snapshot);
    }
    return store;
}
