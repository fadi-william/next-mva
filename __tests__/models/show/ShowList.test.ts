import { getSnapshot } from "mobx-state-tree";
import { getStore, ShowList } from "../../../app/models/show/ShowList";

// Mock the show service.
jest.mock("../../../app/service/showService");

const shows = {
    shows: [{
        id: 757,
        image: "https://img.url/1",
        name: "A show",
        summary: "This is the show summary",
    }, {
        id: 1146,
        image: "https://img.url/2",
        name: "A second show",
        summary: "This is the second show summary",
    }],
};

describe("show list model tests", () => {

    it("can create a show list model", () => {
        const showListStore = ShowList.create(shows);

        expect(showListStore.shows[0].id).toBe(757);
        expect(showListStore.isLoading).toBe(true);
        expect(getSnapshot(showListStore)).toMatchSnapshot();
    });

    it("can get total number of shows", () => {
        const showListStore = ShowList.create();

        expect(showListStore.totalShows).toBe(0);
    });

    it("can load the show list from a web server", async () => {
        const showListStore = ShowList.create();
        await showListStore.afterCreate();
        await showListStore.getShows();

        await new Promise( (res) => setTimeout(() => {
            expect(showListStore.isLoading).toBe(false);
            expect(showListStore.shows[0].id).toBe(975);
            expect(showListStore.shows[0].name).toBe("Batman");
            res();
        }, 3000));
    });

    it("can get a store on the client", () => {
        const showListStore = getStore(false);

        expect(showListStore.isLoading).toBe(true);
    });

    it("can get a store on the server", () => {
        const showListStore = getStore(true, shows);
        expect(showListStore.shows[0].id).toBe(757);
        expect(showListStore.shows[0].name).toBe("A show");
    });

    it("test before destroy lifecycle method", async () => {
        const showListStore = ShowList.create(shows);

        await showListStore.beforeDestroy();

        expect(showListStore.getShows).toThrowError();
    });
});
