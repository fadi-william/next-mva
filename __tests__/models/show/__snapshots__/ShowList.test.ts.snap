// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show list model tests can create a show list model 1`] = `
Object {
  "isLoading": true,
  "shows": Array [
    Object {
      "id": 757,
      "image": "https://img.url/1",
      "isLoading": true,
      "name": "A show",
      "summary": "This is the show summary",
    },
    Object {
      "id": 1146,
      "image": "https://img.url/2",
      "isLoading": true,
      "name": "A second show",
      "summary": "This is the second show summary",
    },
  ],
}
`;
